File: flutter_inappwebview_platform_interface/lib/src/types/web_notification.dart
import 'package:flutter_inappwebview_internal_annotations/flutter_inappwebview_internal_annotations.dart';

import '../web_uri.dart';
import 'text_direction_kind.dart';
import 'enum_method.dart';

part 'web_notification.g.dart';

///Class that represents the data of a web notification.
///
///This corresponds to the properties of the [ICoreWebView2Notification](https://learn.microsoft.com/en-us/microsoft-edge/webview2/reference/win32/icorewebview2notification?view=webview2-1.0.3595.46) interface.
@ExchangeableObject()
class WebNotification_ {
  ///The title of the notification.
  String? title;

  ///The body text of the notification.
  String? body;

  ///The text direction.
  TextDirectionKind_? direction;

  ///The language of the notification.
  String? language;

  ///The tag of the notification.
  String? tag;

  ///The icon URI.
  WebUri? iconUri;

  ///The badge URI.
  WebUri? badgeUri;

  ///The body image URI.
  WebUri? bodyImageUri;

  ///Whether the notification should renotify.
  bool? shouldRenotify;

  ///Whether the notification requires interaction.
  bool? requiresInteraction;

  ///Whether the notification is silent.
  bool? isSilent;

  ///The timestamp of the notification, in milliseconds since UNIX epoch.
  double? timestamp;

  ///The vibration pattern for devices with vibration hardware to emit.
  ///
  ///The vibration pattern can be represented by an array of 64-bit unsigned integers
  ///describing a pattern of vibrations and pauses. See [Vibration API](https://developer.mozilla.org/docs/Web/API/Vibration_API) for more information.
  ///This corresponds to [Notification.vibrate](https://developer.mozilla.org/docs/Web/API/Notification/vibrate) DOM API.
  ///An empty list is returned if no vibration patterns are specified.
  List<int>? vibrationPattern;

  WebNotification_({
    this.title,
    this.body,
    this.direction,
    this.language,
    this.tag,
    this.iconUri,
    this.badgeUri,
    this.bodyImageUri,
    this.shouldRenotify,
    this.requiresInteraction,
    this.isSilent,
    this.timestamp,
    this.vibrationPattern,
  });
}

File: flutter_inappwebview_windows/windows/types/web_notification.cpp
#include "web_notification.h"

namespace flutter_inappwebview_plugin
{
  WebNotification::WebNotification(const std::optional<std::string>& title,
    const std::optional<std::string>& body,
    const std::optional<TextDirectionKind>& direction,
    const std::optional<std::string>& language,
    const std::optional<std::string>& tag,
    const std::optional<std::string>& iconUri,
    const std::optional<std::string>& badgeUri,
    const std::optional<std::string>& bodyImageUri,
    const std::optional<bool>& shouldRenotify,
    const std::optional<bool>& requiresInteraction,
    const std::optional<bool>& isSilent,
    const std::optional<double>& timestamp,
    const std::optional<std::vector<int64_t>>& vibrationPattern)
    : title(title), body(body), direction(direction), language(language), tag(tag), iconUri(iconUri), badgeUri(badgeUri),
    bodyImageUri(bodyImageUri), shouldRenotify(shouldRenotify), requiresInteraction(requiresInteraction), isSilent(isSilent),
    timestamp(timestamp), vibrationPattern(vibrationPattern)
  {}

  WebNotification::WebNotification(const flutter::EncodableMap& map)
    : WebNotification(get_optional_fl_map_value<std::string>(map, "title"),
      get_optional_fl_map_value<std::string>(map, "body"),
      TextDirectionKindFromOptionalInteger(get_optional_fl_map_value<int64_t>(map, "direction")),
      get_optional_fl_map_value<std::string>(map, "language"),
      get_optional_fl_map_value<std::string>(map, "tag"),
      get_optional_fl_map_value<std::string>(map, "iconUri"),
      get_optional_fl_map_value<std::string>(map, "badgeUri"),
      get_optional_fl_map_value<std::string>(map, "bodyImageUri"),
      get_optional_fl_map_value<bool>(map, "shouldRenotify"),
      get_optional_fl_map_value<bool>(map, "requiresInteraction"),
      get_optional_fl_map_value<bool>(map, "isSilent"),
      get_optional_fl_map_value<double>(map, "timestamp"),
      get_optional_fl_map_value<std::vector<int64_t>>(map, "vibrationPattern"))
  {}

  flutter::EncodableMap WebNotification::toEncodableMap() const
  {
    return flutter::EncodableMap{
      {"title", make_fl_value(title)},
      {"body", make_fl_value(body)},
      {"direction", make_fl_value(TextDirectionKindToInteger(direction))},
      {"language", make_fl_value(language)},
      {"tag", make_fl_value(tag)},
      {"iconUri", make_fl_value(iconUri)},
      {"badgeUri", make_fl_value(badgeUri)},
      {"bodyImageUri", make_fl_value(bodyImageUri)},
      {"shouldRenotify", make_fl_value(shouldRenotify)},
      {"requiresInteraction", make_fl_value(requiresInteraction)},
      {"isSilent", make_fl_value(isSilent)},
      {"timestamp", make_fl_value(timestamp)},
      {"vibrationPattern", make_fl_value(vibrationPattern)}
    };
  }
}